pkg/controller/permissions/consumerpermissions.go

```diff
@@ -0,0 +1,157 @@
+		if err != nil {
+			return nil, err
+		}
+		result = append(result, consumerPerms)
+	}
+
+	return result, nil
+}
+
+func CheckConsumerPermsAuthorized(ctx context.Context, c kclient.Client, appInstance *v1.AppInstance, consumedPerms []v1.Permissions) ([]v1.Permissions, error) {
+	imageName := appInstance.Status.AppImage.Name
+
+	// I wonder if this is OK here, since at this point, there could already be a new image staged
+	// and thus the annotation would already be updated and may not match the image that we use here.
+	// That should be a very rare case though and has to have a very good timing.
+	// Anyway, this couldn't be used for privilege escalation, since we're still using the current digest.
+	if oi, ok := appInstance.Annotations[labels.AcornOriginalImage]; ok {
+		imageName = oi
+	}
+
+	authzPerms, err := imagerules.GetAuthorizedPermissions(ctx, c, appInstance.Namespace, imageName, appInstance.Status.AppImage.Digest)
+	if err != nil {
+		return nil, err
+	}
+
+	// Need to deepcopy here since otherwise we'd override the name in the original object which we still need
+	copyWithName := func(perms []v1.Permissions, name string) []v1.Permissions {
+		nperms := make([]v1.Permissions, len(perms))
+		for i := range perms {
+			nperms[i] = perms[i].DeepCopy().Get()
+			nperms[i].ServiceName = name
+		}
+		return nperms
+	}
+
+	denied := []v1.Permissions{}
+
+	for _, tmp := range consumedPerms {
+		if d, granted := v1.GrantsAll(appInstance.Status.Namespace, []v1.Permissions{tmp}, copyWithName(authzPerms, tmp.ServiceName)); !granted {
+			denied = append(denied, d...)
+		}
+	}
+
+	return denied, nil
+}
+
+func ConsumerPermissions(req router.Request, resp router.Response) error {
+	app := req.Object.(*v1.AppInstance)
+
+	iraEnabled, err := config.GetFeature(req.Ctx, req.Client, profiles.FeatureImageRoleAuthorizations)
+	if err != nil {
+		return err
+	}
+	if !iraEnabled {
+		// Reset status field if IRAs are disabled
+		app.Status.DeniedConsumerPermissions = nil
+	}
+
+	// Register for re-triggering when any service instance changes
+	svclist := &v1.ServiceInstanceList{}
+	if err := req.List(svclist, &kclient.ListOptions{
+		Namespace: app.Status.Namespace,
+	}); err != nil {
+		return err
+	}
+
+	consumerPerms, err := collectConsumerPermissions(req.Ctx, req.Client, app)
+	if err != nil {
+		return err
+	}
+
+	newPerms, ok := v1.GrantsAll(app.Namespace, consumerPerms, app.Status.Permissions)
+	if ok {
+		// nothing new -> nothing to do
+		app.Status.DeniedConsumerPermissions = nil
+		return nil
+	}
+
+	if iraEnabled {
+		denied, err := CheckConsumerPermsAuthorized(req.Ctx, req.Client, app, newPerms)
+		if err != nil {
+			return err
+		}
+
+		if len(denied) > 0 {
+			app.Status.DeniedConsumerPermissions = denied
+			return nil
+		}
+	}
+
+	app.Status.DeniedConsumerPermissions = nil
+	app.Status.Permissions = v1.SimplifySet(append(app.Status.Permissions, newPerms...))
+	return nil
```

Comment:
In a few spots here, the app status is modified and then we `return nil`. Don't we need to pass this to `resp.Objects` or update it with the client or something in order for the changes to stick?

Reply:
They just stick without anything else to be done :grin: 
I'm not knees deep in the details there, as that's just what I did before without questioning it, but I assume that the request object is always part of the response :+1: